### PR TITLE
Force in preview as this can cause problems in legacy scenarios…

### PR DESCRIPTION
**Pull request by nkolev92**
… where a project file was altered outside of VS to remove a package reference and then the same package reference was added in the package manager

Fixes https://github.com/NuGet/Home/issues/5542

This is kind of hard to add automated tests for with the current infrastructure. 
Fastest/Easiest way is to ask the vendors to start doing manual tests for now. 
Tracking issue here for creating the automated tests https://github.com/NuGet/Home/issues/5543

### CODE REVIEW

**Review by emgarten — APPROVED** [review]
LGTM, I can't think of a scenario where preview shouldn't force.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/NoOpRestoreResult.cs
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/NoOpRestoreResult.cs:19]: how does passing  null vs empty list make a difference?

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs:144]: just curious, wouldn't there be major perf loss too with no-op now? can we scope this down for legacy scenarios?

Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs:144]: Preview is used only on install. So the no-op scenario should've never been hit, except for the above corner case.
No perf concerns here.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/NoOpRestoreResult.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/NoOpRestoreResult.cs:19]: We try to do a linq operation on that value, so NPE happens (this should never happen now, since we force)

**Review by alpaix — APPROVED** [review]

**Review by jainaashish — APPROVED** [review]

**Comment by nkolev92** [comment]
Added an end-to-end test.
Will wait for the test to pass on the CI and then merge.

**Comment by nkolev92** [comment]
The test is passing
Can one of you @alpaix @jainaashish @emgarten quickly peek the test as a sanity check, before I merge.

**Comment by emgarten** [comment]
The test looks good to me

File: test/EndToEnd/tests/BuildIntegratedTest.ps1
Comment (alpaix) [inline test/EndToEnd/tests/BuildIntegratedTest.ps1:518]: Act should refer to a single action

File: test/EndToEnd/tests/BuildIntegratedTest.ps1
Comment (alpaix) [inline test/EndToEnd/tests/BuildIntegratedTest.ps1:512]: duplicate

File: test/EndToEnd/tests/BuildIntegratedTest.ps1
Comment (alpaix) [inline test/EndToEnd/tests/BuildIntegratedTest.ps1:502]: remove comments

File: test/EndToEnd/tests/BuildIntegratedTest.ps1
Comment (alpaix) [inline test/EndToEnd/tests/BuildIntegratedTest.ps1:515]: remove or make it meaningful

File: test/EndToEnd/tests/BuildIntegratedTest.ps1
Comment (alpaix) [inline test/EndToEnd/tests/BuildIntegratedTest.ps1:520]: remove or make the output message meaningful

File: test/EndToEnd/tests/BuildIntegratedTest.ps1
Comment (alpaix) [inline test/EndToEnd/tests/BuildIntegratedTest.ps1:519]: Does this remove references from a file?

File: test/EndToEnd/tests/BuildIntegratedTest.ps1
Comment (alpaix) [inline test/EndToEnd/tests/BuildIntegratedTest.ps1:544]: remove

File: test/EndToEnd/tests/BuildIntegratedTest.ps1
Comment (alpaix) [inline test/EndToEnd/tests/BuildIntegratedTest.ps1:530]: => `Remove-AllPackageReferences`

File: test/EndToEnd/tests/BuildIntegratedTest.ps1
Comment (alpaix) [inline test/EndToEnd/tests/BuildIntegratedTest.ps1:536]: remove

File: test/EndToEnd/tests/BuildIntegratedTest.ps1
Comment (alpaix) [inline test/EndToEnd/tests/BuildIntegratedTest.ps1:543]: If you want to remove particular reference then provide a parameter

File: test/EndToEnd/tests/BuildIntegratedTest.ps1
Comment (alpaix) [inline test/EndToEnd/tests/BuildIntegratedTest.ps1:524]: I presume package should be uninstalled by this moment? Add a precondition?

**Comment by alpaix** [comment]
@nkolev92 Sorry I couldn't review the test within 3 minutes before you merged. The test needs to be cleaned and clarified.

**Comment by nkolev92** [comment]
@alpaix 
I was maybe too quick on the trigger. :) 
I will clean it up and only do it on dev?
I'd rather not add more noise to the preview4 branch since the changes are not functional.
Do you agree?

**Comment by alpaix** [comment]
Was it approved for preview4?

**Comment by nkolev92** [comment]
Yes. 
We don't want install to be broken.

**Pull request merged**
